tools/ci-cdk/canary-runner/src/run.rs

```diff
@@ -62,23 +63,82 @@ pub struct RunOpt {
     #[clap(long)]
     musl: bool,
 
-    /// The name of the S3 bucket to upload the canary binary bundle to
+    /// File path to a CDK outputs JSON file. This can be used instead
+    /// of all the --lambda... args.
     #[clap(long)]
-    lambda_code_s3_bucket_name: String,
+    cdk_output: Option<PathBuf>,
+
+    /// The name of the S3 bucket to upload the canary binary bundle to
+    #[clap(long, required_unless_present = "cdk-output")]
+    lambda_code_s3_bucket_name: Option<String>,
 
     /// The name of the S3 bucket for the canary Lambda to interact with
-    #[clap(long)]
-    lambda_test_s3_bucket_name: String,
+    #[clap(long, required_unless_present = "cdk-output")]
+    lambda_test_s3_bucket_name: Option<String>,
 
     /// The ARN of the role that the Lambda will execute as
-    #[clap(long)]
+    #[clap(long, required_unless_present = "cdk-output")]
+    lambda_execution_role_arn: Option<String>,
+}
+
+#[derive(Debug)]
+struct Options {
+    sdk_version: Option<String>,
+    sdk_path: Option<PathBuf>,
+    musl: bool,
+    lambda_code_s3_bucket_name: String,
+    lambda_test_s3_bucket_name: String,
     lambda_execution_role_arn: String,
 }
 
+impl Options {
+    fn load_from(run_opt: RunOpt) -> Result<Options> {
+        if let Some(cdk_output) = &run_opt.cdk_output {
+            #[derive(Deserialize)]
+            struct Inner {
+                #[serde(rename = "canarycodebucketname")]
+                lambda_code_s3_bucket_name: String,
+                #[serde(rename = "canarytestbucketname")]
+                lambda_test_s3_bucket_name: String,
+                #[serde(rename = "lambdaexecutionrolearn")]
+                lambda_execution_role_arn: String,
+            }
+            #[derive(Deserialize)]
+            struct Outer {
+                #[serde(rename = "aws-sdk-rust-canary-stack")]
+                inner: Inner,
+            }
+
+            let value: Outer = serde_json::from_reader(
+                std::fs::File::open(cdk_output).context("open cdk output")?,
+            )
+            .context("read cdk output")?;
+            Ok(Options {
+                sdk_version: run_opt.sdk_version,
+                sdk_path: run_opt.sdk_path,
+                musl: run_opt.musl,
+                lambda_code_s3_bucket_name: value.inner.lambda_code_s3_bucket_name,
+                lambda_test_s3_bucket_name: value.inner.lambda_test_s3_bucket_name,
+                lambda_execution_role_arn: value.inner.lambda_execution_role_arn,
+            })
+        } else {
+            Ok(Options {
+                sdk_version: run_opt.sdk_version,
+                sdk_path: run_opt.sdk_path,
+                musl: run_opt.musl,
+                lambda_code_s3_bucket_name: run_opt.lambda_code_s3_bucket_name.expect("required"),
+                lambda_test_s3_bucket_name: run_opt.lambda_test_s3_bucket_name.expect("required"),
+                lambda_execution_role_arn: run_opt.lambda_execution_role_arn.expect("required"),
```

Comment:
I feel like we should be using `clap` to do this required check. I don't have strong feelings on it though.

Reply:
Clap is doing the required check already. I have to unwrap the options to get to the data in this case (but it's already validated).